File: clients/typescript/src/dtos.ts
export interface Message {
    id: string;
    previous?: Message | null;
    type: string;
    createdAt: string;
    appName?: string;
}

export interface ActionRequest<T> extends Message {
    route: string;
    payload?: T;
}

export interface ActionError {
    type: string;
    message: string;
}

export interface ActionResponse<T> extends Message {
    success: boolean;
    logs: string[];
    result?: T;
    error?: ActionError | null;
}

export interface TopicMessage<T> extends Message {
    topic: string;
    payload?: T;
}

export interface Resources {
    actions: string[];
    caches: string[];
    topics: string[];
    version: string;
}

export interface CacheItem<T> {
    group: string;
    key: string;
    value?: T;
}

export interface CacheEvent<T> {
    group: string;
    key: string;
    value?: T;
}

File: clients/csharp/AmenoLink/Dtos.cs
namespace AmenoLink;
 
public record Message(
    string Id,
    string CreatedAt,
    string Type,
    Message? Previous = null,
    string? AppName = null
);

public record ActionRequest<T>(
    string Id,
    string CreatedAt,
    string Route,
    string Type = "ActionRequest",
    T? Payload = default,
    Message? Previous = null,
    string? AppName = null
) : Message(Id, CreatedAt, Type, Previous, AppName);

public record ActionError(
    string Type,
    string Message
);

public record ActionResponse<T>(
    string Id,
    string CreatedAt,
    bool Success,
    string[] Logs,
    string Type = "ActionResponse",
    T? Result = default,
    ActionError? Error = null,
    Message? Previous = null,
    string? AppName = null
) : Message(Id, CreatedAt, Type, Previous, AppName);

public record TopicMessage<T>(
    string Id,
    string CreatedAt,
    string Topic,
    string Type = "TopicMessage",
    T? Payload = default,
    Message? Previous = null,
    string? AppName = null
) : Message(Id, CreatedAt, Type, Previous, AppName);

public record Resources(
    string[] Actions,
    string[] Caches,
    string[] Topics,
    string Version
);

public record CacheItem<T>(
    string Group,
    string Key,
    T? Value = default
);

public record CacheEvent<T>(
    string Group,
    string Key,
    T? Value = default
);
